Alerting: Configure alert manager data source as an external AM

konrad147 opened the pull request:
**What this PR does / why we need it**:
This PR enables an alert manager data source to be set as an external alert manager for Grafana-managed alerts

**Which issue(s) this PR fixes**:
Fixes #51424

**Special notes for your reviewer**:
This is the first part of the required work. As a next step, we would like to get rid of the URL-based external alert managers

**PREVIEW**
![image](https://user-images.githubusercontent.com/6293563/180188426-2d86814f-8b92-4561-ab58-944f11a9f099.png)

konrad147 committed:
Add external AM setting as a part of the data source configuration

konrad147 committed:
Add AM dropdown example

konrad147 committed:
Remove AM selector

konrad147 committed:
Add deprecation notice

konrad147 committed:
Display AM data sources when no external AMs exist

konrad147 committed:
Improve external AMs section

konrad147 committed:
Remove external AM config section for non-Grafana Alertmanagers

konrad147 committed:
Display alertmanager url if available

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/51424-external-AM-as-a-datasource

joshhunt commented on .betterer.results:
```diff
@@ -6445,9 +6445,6 @@ exports[`better eslint`] = {
       [0, 0, 0, "Unexpected any. Specify a different type.", "9"],
       [0, 0, 0, "Do not use any type assertions.", "10"]
     ],
-    "public/app/plugins/datasource/alertmanager/ConfigEditor.tsx:5381": [
-      [0, 0, 0, "Do not use any type assertions.", "0"]
-    ],
```
🫡

soniaAguilarPeiron reviewed (approved):
LGTM ✅

gotjosh commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -121,10 +136,43 @@ export const ExternalAlertmanagers = () => {
     <div>
       <h4>External Alertmanagers</h4>
+
+      <h5>Alert managers data sources</h5>
```
```suggestion
      <h5>Alertmanagers data sources</h5>
```

gotjosh commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -121,10 +136,43 @@ export const ExternalAlertmanagers = () => {
+
+      <h5>Alert managers data sources</h5>
+      <div className={styles.muted}>
+        During the Alert manager data source configuration you can choose to send Grafana managed alerts to that Alert
+        manager. <br />
+        Here you can see the list of all Alert manager data sources which receive Grafana managed alerts.
```
```suggestion
        Alertmanager data sources support a configuration setting that allows you to choose to send Grafana-managed alerts to that Alertmanager. <br />
        Below, you can see the list of all Alertmanager data sources that have this setting enabled.
```

gotjosh commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -121,10 +136,43 @@ export const ExternalAlertmanagers = () => {
+        ))}
+      </div>
+
+      <h5>URL-based external alert managers [deprecated]</h5>
```
```suggestion
      <h5>Alertmanagers by URL</h5>
```

gotjosh commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -121,10 +136,43 @@ export const ExternalAlertmanagers = () => {
+      </div>
+
+      <h5>URL-based external alert managers [deprecated]</h5>
+      <Alert severity="warning" title="Deprecation notice">
```
```suggestion
      <Alert severity="warning" title="Deprecation Notice">
```

gotjosh commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -121,10 +136,43 @@ export const ExternalAlertmanagers = () => {
+
+      <h5>URL-based external alert managers [deprecated]</h5>
+      <Alert severity="warning" title="Deprecation notice">
+        The URL-based configuration method is deprecated and will be removed in the future versions of Grafana <br />
+        Please use Alertmanager data sources to configure external alert managers.
```
```suggestion
        The URL-based configuration of Alertmanagers is now deprecated and will be removed in Grafana 9.2.0<br />
        Please use Alertmanager data sources to configure your external Alertmanagers.
```

gotjosh commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -188,20 +237,24 @@ export const ExternalAlertmanagers = () => {
+        <div className={styles.amChoice}>
+          <Field
+            label="Send alerts to"
+            description="Sets which Alertmanager will handle your alerts. Internal (Grafana built in Alertmanager), External (All Alertmanagers configured above), or both."
```
```suggestion
            description="Configures how the Grafana alert rule evaluation engine Alertmanager will handle your alerts. Internal (Grafana built-in Alertmanager), External (All Alertmanagers configured above), or both."
```

gotjosh commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -188,20 +237,24 @@ export const ExternalAlertmanagers = () => {
       )}
+
+      {hasExternalAlertmanagers && (
+        <div className={styles.amChoice}>
```
The way that the elements are arranged it makes it unclear whether this option is going to be removed or not - I'd consider moving this section right after the "Alertmanager data sources" section to make it a bit more clear.

gotjosh commented on public/app/plugins/datasource/alertmanager/ConfigEditor.tsx:
```diff
@@ -48,13 +49,31 @@ export const ConfigEditor = (props: Props) => {
         </div>
+        <div className="gf-form-inline">
+          <InlineField
+            label="Handle Grafana alerts"
```
```suggestion
            label="Receive Grafana-managed alerts"
```

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/51424-external-AM-as-a-datasource

gotjosh commented on public/app/plugins/datasource/alertmanager/ConfigEditor.tsx:
```diff
@@ -48,13 +49,31 @@ export const ConfigEditor = (props: Props) => {
+        <div className="gf-form-inline">
+          <InlineField
+            label="Handle Grafana alerts"
+            tooltip="When enabled Grafana Managed alerts will be sent to this Alertmanager"
```
```suggestion
            tooltip="When enabled, Grafana-managed alerts are sent to this Alertmanager."
```

gotjosh commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -121,10 +136,43 @@ export const ExternalAlertmanagers = () => {
+      </div>
+      <div className={styles.externalDs}>
+        {externalDsAlertManagers.map((ds) => (
+          <Card key={ds.uid}>
```
Can we make this a link back to the datasource? It would be useful to be able to link back.

gotjosh reviewed (commented):
LGTM, thank you very much for your work, Konrad.

Can you please take a look at my comments?

gotjosh committed:
Re-use the same parsing logic as the datasource

As we attempt to parse the datasource URL before we pass it to the sender, let's use the same parsing logic as the datasource uses to make sure it matches what the customer sees.

We found out about this because we used an URL with no scheme that passed the health check but was invalid at the moment of receiving alerts.

Co-Authored-By: Jean-Philippe Quéméner <JohnnyQQQQ@users.noreply.github.com>

gotjosh commented:
Two more things:

1. The previous logic used the `/alertmanagers` endpoint to indicate you that your new external alertmanagers is now connected - we should probably do the same thing.
2. @JohnnyQQQQ and I just pushed a small fixed to make sure we parse URLs in the same way that the datasource does as we connect our alerting engine with the external Alertmanager.

konrad147 committed:
Improve user-facing messages

Co-authored-by: gotjosh <josue.abreu@gmail.com>

konrad147 committed:
Add data sources URLs

konrad147 committed:
Add displaying of external AMs statuses

konrad147 commented:
Improved version
![image](https://user-images.githubusercontent.com/6293563/180463131-2711c789-2b43-44e0-be60-33a6e964faae.png)

gotjosh committed:
Parse the URL even if there's no basic auth.

gotjosh reviewed (commented):
Nice work @konrad147 we're almost there!

I noticed that the whole time I've been testing with a datasource that doesn't include the scheme (e.g. localhost:9093) but passes all the health checks. 

When you use a datasource with no scheme, despite working as expected and having alerts delivered it ends up with a yellow heart indefinetly. This is because Grafana prepends the scheme if it's not present. 

Could you maybe do a substring matching on the URL?

konrad147 committed:
Improved external AMs

konrad147 committed:
Merge branch 'alerting/51424-external-AM-as-a-datasource' of github.com:grafana/grafana into alerting/51424-external-AM-as-a-datasource

konrad147 commented:
![image](https://user-images.githubusercontent.com/6293563/180952022-afefdb8e-9763-41da-8eee-c1f8c8f0c467.png)

JohnnyQQQQ commented:
That's amazing

konrad147 committed:
Add a warning when multiple AMs have the same url

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/51424-external-AM-as-a-datasource

JohnnyQQQQ commented:
I did another pass, the problem that Josh did mention is still present. When using `localhost:9093` as the datasource URL without `http://` or `https://` in front of it, it will stay in `pending` forever.

konrad147 commented on public/app/features/alerting/unified/hooks/useExternalAmSelector.ts:
```diff
@@ -51,3 +58,67 @@ export function useExternalAmSelector(): AlertmanagerConfig[] | [] {
+    const isDropped = matchingDroppedUrls > 0;
+    const isActive = matchingActiveUrls > 0;
+
+    const isStatusInconclusive = matchingDroppedUrls > 1 || matchingActiveUrls > 1;
```
If someone configures 2 data sources but e.g. with different credentials. The alertmanager response contains only URLs so we are not able to say for sure that this is actually the status of this particular data source

gillesdemey replied:
That makes total sense, can we add that as code comments so we can remember that rationale next time?

konrad147 commented on public/app/features/alerting/unified/hooks/useExternalAmSelector.ts:
```diff
@@ -51,3 +58,67 @@ export function useExternalAmSelector(): AlertmanagerConfig[] | [] {
+function getDataSourceUrlWithProtocol<T>(dsSettings: DataSourceSettings<T>) {
+  const hasProtocol = new RegExp('^[^:]*://').test(dsSettings.url);
+  if (!hasProtocol) {
+    return `http://${dsSettings.url}`; // Grafana append http protocol if there is no any
```
Well, this is kind of a workaround for the poor validation of the data source configuration 😕 
You can configure a datasource of e.g. a `localhost:9093` URL (without specifying the protocol) and everything will work fine. However, the alertmanager response will add `http` prefix to URLs that don't have them. So to match the URL we also need to add the protocol.
I added `http` as it's the same what is attached on the backend

gillesdemey replied:
Ah I suspected something along those lines, does make me wonder why they chose `http` over `https` but that's not a concern related to this PR 👍

gillesdemey commented:
One last general comment (but not a blocker to merge this) is we should perhaps add an empty state for data sources explaining that you don't have any configured and can add an external AM by enabling that toggle in an Alert Manager data source

konrad147 committed:
Handle only internal state of alertmanager configuration

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/51424-external-AM-as-a-datasource

konrad147 committed:
Improve no alertmanagers handling, add more explanation of the changes

konrad147 committed:
Refactor components

konrad147 committed:
Add tests

konrad147 committed:
Fix dropped status, add more tests

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/51424-external-AM-as-a-datasource

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/51424-external-AM-as-a-datasource

brendamuir commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagerDataSources.tsx:
```diff
@@ -0,0 +1,122 @@
+        <CallToActionCard
+          message={
+            <div>
+              There are no Alertmanager data sources configured to receive Grafana-managed alerts <br />
+              You can change this by selecting Receive Grafana Alerts in a data source configuration
+            </div>
```
```suggestion
              There are no Alertmanager data sources configured to receive Grafana-managed alerts. <br />
              You can change this by selecting Receive Grafana Alerts in a data source configuration.
            </div>
```

brendamuir commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagerDataSources.tsx:
```diff
@@ -0,0 +1,122 @@
+      <Card.Heading className={styles.externalHeading}>
+        {dataSource.name}{' '}
+        {statusInconclusive && (
+          <Tooltip content="Multiple Alertmangers have the same URL configured. The state might be inconclusive">
+            <Icon name="exclamation-triangle" size="md" className={styles.externalWarningIcon} />
+          </Tooltip>
```
```suggestion
          <Tooltip content="Multiple Alertmangers have the same URL configured. The state might be inconclusive.">
            <Icon name="exclamation-triangle" size="md" className={styles.externalWarningIcon} />
          </Tooltip>
```

brendamuir commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagerDataSources.tsx:
```diff
@@ -0,0 +1,122 @@
+          <Badge
+            text="Inactive"
+            color="red"
+            tooltip="Grafana is configured to send alerts to the built-in internal alermanager only. External Alertmanages will not receive any alerts"
+          />
```
```suggestion
            tooltip="Grafana is configured to send alerts to the built-in internal Alertmanager only. External Alertmanagers do not receive any alerts."
          />
```

brendamuir commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -121,10 +136,47 @@ export const ExternalAlertmanagers = () => {
   return (
     <div>
       <h4>External Alertmanagers</h4>
+      <Alert title="External alertmanagers changes" severity="info">
+        We changed the way of management of external Alertmanagers
+        <br />
+        From now on, you can use configured Alertmanager data sources as receivers of your Grafana-managed alerts
+        <br />
+        Check our documentation to find our more.
```
```suggestion
      <Alert title="External Alertmanager changes" severity="info">
        The way you configure external Alertmanagers has changed. 
        <br />
        You can now use configured Alertmanager data sources as receivers of your Grafana-managed alerts.
        <br />
        For more information, refer to our documentation.
```

brendamuir commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -121,10 +136,47 @@ export const ExternalAlertmanagers = () => {
+      {hasExternalAlertmanagers && (
+        <div className={styles.amChoice}>
+          <Field
+            label="Send alerts to"
+            description="Configures how the Grafana alert rule evaluation engine Alertmanager will handle your alerts. Internal (Grafana built-in Alertmanager), External (All Alertmanagers configured above), or both."
```
```suggestion
            label="Send alerts to"
            description="Configures how the Grafana alert rule evaluation engine Alertmanager handles your alerts. Internal (Grafana built-in Alertmanager), External (All Alertmanagers configured above), or both."
```

brendamuir commented on public/app/features/alerting/unified/components/admin/ExternalAlertmanagers.tsx:
```diff
@@ -121,10 +136,47 @@ export const ExternalAlertmanagers = () => {
+
+      <h5>Alertmanagers by URL</h5>
+      <Alert severity="warning" title="Deprecation Notice">
+        The URL-based configuration of Alertmanagers is now deprecated and will be removed in Grafana 9.2.0
+        <br />
+        Please use Alertmanager data sources to configure your external Alertmanagers.
```
```suggestion
        The URL-based configuration of Alertmanagers is deprecated and will be removed in Grafana 9.2.0.
        <br />
        Use Alertmanager data sources to configure your external Alertmanagers.
```

brendamuir commented on public/app/features/alerting/unified/hooks/useExternalAmSelector.ts:
```diff
@@ -51,3 +58,71 @@ export function useExternalAmSelector(): AlertmanagerConfig[] | [] {
+    const isDropped = matchingDroppedUrls > 0;
+    const isActive = matchingActiveUrls > 0;
+
+    // It may happen that there are multiple Alertmanagers of the same URL (e.g. only different credentials)
+    // Alertmanager response contains only URLs so in case of duplication we are not able
+    // to distinguish which is which, hence the status is inconclusive
+    const isStatusInconclusive = matchingDroppedUrls + matchingActiveUrls > 1;
```
```suggestion
    // Multiple Alertmanagers of the same URL may exist (e.g. with different credentials)
    // Alertmanager response only contains URLs, so in case of duplication, we are not able
    // to distinguish which is which, resulting in an inconclusive status. 
    const isStatusInconclusive = matchingDroppedUrls + matchingActiveUrls > 1;
```

konrad147 committed:
Apply suggestions from code review

Improve wording

Co-authored-by: brendamuir <100768211+brendamuir@users.noreply.github.com>

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/51424-external-AM-as-a-datasource

konrad147 committed:
Fix prettier

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/51424-external-AM-as-a-datasource

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/51424-external-AM-as-a-datasource
